# Demonstration of the Data Processing file

This file will go through the reasoning behind the preprocessing done in utils/data_preprocessing

## Setup

We import the DataProcesser, and configure polars for the demonstration

In [1]:
import os
import sys
import polars as pl
import numpy as np

parent_dir = os.path.abspath(os.path.join(os.getcwd(), "../utils"))
sys.path.append(parent_dir)

from data_preprocessing import DataProcesser
dataProcesser = DataProcesser()

pl.Config.set_tbl_cols(-1)

polars.config.Config

The DataProcesser contains the imported dataframes as local variables, so that we don't have to reimport them for every processing

In [2]:
print("Articles shape:        ", dataProcesser.articles_df.shape)
print("Document vectors shape:", dataProcesser.articles_df.shape)
print("Train Behaviours shape:", dataProcesser.articles_df.shape)
print("Test Behaviours shape: ", dataProcesser.articles_df.shape)

Articles shape:         (20738, 21)
Document vectors shape: (20738, 21)
Train Behaviours shape: (20738, 21)
Test Behaviours shape:  (20738, 21)


I will no go trough the processing done to each of the dataframes (with the Behaviours being combined into one)

## Preprocessing

### Articles DataFrame

#### Exploration

This DataFrame contains information about the different articles used in the dataframe

In [3]:
dataProcesser.articles_df.head(1)

shape: (1, 21)
┌─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┐
│ art ┆ tit ┆ sub ┆ las ┆ pre ┆ bod ┆ pub ┆ ima ┆ art ┆ url ┆ ner ┆ ent ┆ top ┆ cat ┆ sub ┆ cat ┆ tot ┆ tot ┆ tot ┆ sen ┆ sen │
│ icl ┆ le  ┆ tit ┆ t_m ┆ miu ┆ y   ┆ lis ┆ ge_ ┆ icl ┆ --- ┆ _cl ┆ ity ┆ ics ┆ ego ┆ cat ┆ ego ┆ al_ ┆ al_ ┆ al_ ┆ tim ┆ tim │
│ e_i ┆ --- ┆ le  ┆ odi ┆ m   ┆ --- ┆ hed ┆ ids ┆ e_t ┆ str ┆ ust ┆ _gr ┆ --- ┆ ry  ┆ ego ┆ ry_ ┆ inv ┆ pag ┆ rea ┆ ent ┆ ent │
│ d   ┆ str ┆ --- ┆ fie ┆ --- ┆ str ┆ _ti ┆ --- ┆ ype ┆     ┆ ers ┆ oup ┆ lis ┆ --- ┆ ry  ┆ str ┆ iew ┆ evi ┆ d_t ┆ _sc ┆ _la │
│ --- ┆     ┆ str ┆ d_t ┆ boo ┆     ┆ me  ┆ lis ┆ --- ┆     ┆ --- ┆ s   ┆ t[s ┆ i16 ┆ --- ┆ --- ┆ s   ┆ ews ┆ ime ┆ ore ┆ bel │
│ i32 ┆     ┆     ┆ ime ┆ l   ┆     ┆ --- ┆ t[i ┆ str ┆     ┆ lis ┆ --- ┆ tr] ┆     ┆ lis ┆ str ┆ --- ┆ --- ┆ --- ┆ --- ┆ --- │
│     ┆     ┆     ┆ --- ┆     ┆     ┆ dat ┆ 64] ┆     ┆     ┆ t[s ┆ lis ┆     ┆     ┆ t[i ┆     ┆ i32 ┆ i32 ┆ f32 ┆ f32 ┆ str │
│     ┆     ┆     ┆ dat ┆     ┆     ┆ eti ┆     ┆     ┆     ┆ tr] ┆ t[s ┆     ┆     ┆ 16] ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆     ┆ eti ┆     ┆     ┆ me[ ┆     ┆     ┆     ┆     ┆ tr] ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆     ┆ me[ ┆     ┆     ┆ μs] ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆     ┆ μs] ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
╞═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╡
│ 300 ┆ Nat ┆ Pol ┆ 202 ┆ fal ┆ Sag ┆ 200 ┆ [31 ┆ art ┆ htt ┆ []  ┆ []  ┆ ["K ┆ 140 ┆ []  ┆ kri ┆ nul ┆ nul ┆ nul ┆ 0.9 ┆ Neg │
│ 135 ┆ asc ┆ iti ┆ 3-0 ┆ se  ┆ en  ┆ 6-0 ┆ 508 ┆ icl ┆ ps: ┆     ┆     ┆ rim ┆     ┆     ┆ mi  ┆ l   ┆ l   ┆ l   ┆ 955 ┆ ati │
│ 3   ┆ ha  ┆ et  ┆ 6-2 ┆     ┆ om  ┆ 8-3 ┆ 50] ┆ e_d ┆ //e ┆     ┆     ┆ ina ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ ve  │
│     ┆ var ┆ fry ┆ 9   ┆     ┆ den ┆ 1   ┆     ┆ efa ┆ kst ┆     ┆     ┆ lit ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆ ikk ┆ gte ┆ 06: ┆     ┆ øst ┆ 08: ┆     ┆ ult ┆ rab ┆     ┆     ┆ et" ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆ e   ┆ r   ┆ 20: ┆     ┆ ris ┆ 06: ┆     ┆     ┆ lad ┆     ┆     ┆ ,   ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆ den ┆ nu, ┆ 33  ┆     ┆ ke  ┆ 45  ┆     ┆     ┆ et. ┆     ┆     ┆ "Pe ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆ før ┆ at  ┆     ┆     ┆ Nat ┆     ┆     ┆     ┆ dk/ ┆     ┆     ┆ rso ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆ ste ┆ Nat ┆     ┆     ┆ asc ┆     ┆     ┆     ┆ kri ┆     ┆     ┆ nfa ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆ asc ┆     ┆     ┆ ha… ┆     ┆     ┆     ┆ mi/ ┆     ┆     ┆ rli ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆ …   ┆     ┆     ┆     ┆     ┆     ┆     ┆ …   ┆     ┆     ┆ g…  ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
└─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┘

By inspecting null values, we see that of 20738, are around half of inviews, pageviews and readtime rows are empty. Furthermore around 1900 image_ids (~10%) are missing.

In [4]:
print("Dataframe Shape = ", dataProcesser.articles_df.shape)
dataProcesser.articles_df.null_count()

Dataframe Shape =  (20738, 21)


shape: (1, 21)
┌─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┐
│ art ┆ tit ┆ sub ┆ las ┆ pre ┆ bod ┆ pub ┆ ima ┆ art ┆ url ┆ ner ┆ ent ┆ top ┆ cat ┆ sub ┆ cat ┆ tot ┆ tot ┆ tot ┆ sen ┆ sen │
│ icl ┆ le  ┆ tit ┆ t_m ┆ miu ┆ y   ┆ lis ┆ ge_ ┆ icl ┆ --- ┆ _cl ┆ ity ┆ ics ┆ ego ┆ cat ┆ ego ┆ al_ ┆ al_ ┆ al_ ┆ tim ┆ tim │
│ e_i ┆ --- ┆ le  ┆ odi ┆ m   ┆ --- ┆ hed ┆ ids ┆ e_t ┆ u32 ┆ ust ┆ _gr ┆ --- ┆ ry  ┆ ego ┆ ry_ ┆ inv ┆ pag ┆ rea ┆ ent ┆ ent │
│ d   ┆ u32 ┆ --- ┆ fie ┆ --- ┆ u32 ┆ _ti ┆ --- ┆ ype ┆     ┆ ers ┆ oup ┆ u32 ┆ --- ┆ ry  ┆ str ┆ iew ┆ evi ┆ d_t ┆ _sc ┆ _la │
│ --- ┆     ┆ u32 ┆ d_t ┆ u32 ┆     ┆ me  ┆ u32 ┆ --- ┆     ┆ --- ┆ s   ┆     ┆ u32 ┆ --- ┆ --- ┆ s   ┆ ews ┆ ime ┆ ore ┆ bel │
│ u32 ┆     ┆     ┆ ime ┆     ┆     ┆ --- ┆     ┆ u32 ┆     ┆ u32 ┆ --- ┆     ┆     ┆ u32 ┆ u32 ┆ --- ┆ --- ┆ --- ┆ --- ┆ --- │
│     ┆     ┆     ┆ --- ┆     ┆     ┆ u32 ┆     ┆     ┆     ┆     ┆ u32 ┆     ┆     ┆     ┆     ┆ u32 ┆ u32 ┆ u32 ┆ u32 ┆ u32 │
│     ┆     ┆     ┆ u32 ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
╞═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╡
│ 0   ┆ 0   ┆ 0   ┆ 0   ┆ 0   ┆ 0   ┆ 0   ┆ 187 ┆ 0   ┆ 0   ┆ 0   ┆ 0   ┆ 0   ┆ 0   ┆ 0   ┆ 0   ┆ 107 ┆ 108 ┆ 108 ┆ 0   ┆ 0   │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆ 8   ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ 70  ┆ 82  ┆ 82  ┆     ┆     │
└─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┘

#### Summary

**Filtering:** We can remove the user interactions from the DataFrame, as these values can be found by using the Behaviours table. Furthermore, the image_id's shouldn't be important, so these can be removed too. 

**Sorting:** We can sort by published_time if we want to. But this isn't required

**Summary:** Filter out ["total_inviews", "total_pageviews", "total_read_time", "image_ids"], sort by "published_time"

In [5]:
droppable_columns = ["total_inviews", "total_pageviews", "total_read_time", "image_ids"]
dataProcesser.process_dataframe(df = dataProcesser.articles_df, remove_columns = droppable_columns, sort_by = "published_time")

shape: (20_738, 17)
┌─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┐
│ art ┆ tit ┆ sub ┆ las ┆ pre ┆ bod ┆ pub ┆ art ┆ url ┆ ner ┆ ent ┆ top ┆ cat ┆ sub ┆ cat ┆ sen ┆ sen │
│ icl ┆ le  ┆ tit ┆ t_m ┆ miu ┆ y   ┆ lis ┆ icl ┆ --- ┆ _cl ┆ ity ┆ ics ┆ ego ┆ cat ┆ ego ┆ tim ┆ tim │
│ e_i ┆ --- ┆ le  ┆ odi ┆ m   ┆ --- ┆ hed ┆ e_t ┆ str ┆ ust ┆ _gr ┆ --- ┆ ry  ┆ ego ┆ ry_ ┆ ent ┆ ent │
│ d   ┆ str ┆ --- ┆ fie ┆ --- ┆ str ┆ _ti ┆ ype ┆     ┆ ers ┆ oup ┆ lis ┆ --- ┆ ry  ┆ str ┆ _sc ┆ _la │
│ --- ┆     ┆ str ┆ d_t ┆ boo ┆     ┆ me  ┆ --- ┆     ┆ --- ┆ s   ┆ t[s ┆ i16 ┆ --- ┆ --- ┆ ore ┆ bel │
│ i32 ┆     ┆     ┆ ime ┆ l   ┆     ┆ --- ┆ str ┆     ┆ lis ┆ --- ┆ tr] ┆     ┆ lis ┆ str ┆ --- ┆ --- │
│     ┆     ┆     ┆ --- ┆     ┆     ┆ dat ┆     ┆     ┆ t[s ┆ lis ┆     ┆     ┆ t[i ┆     ┆ f32 ┆ str │
│     ┆     ┆     ┆ dat ┆     ┆     ┆ eti ┆     ┆     ┆ tr] ┆ t[s ┆     ┆     ┆ 16] ┆     ┆     ┆     │
│     ┆     ┆     ┆ eti ┆     ┆     ┆ me[ ┆     ┆     ┆     ┆ tr] ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆     ┆ me[ ┆     ┆     ┆ μs] ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆     ┆ μs] ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
╞═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╡
│ 980 ┆ Pan ┆ Han ┆ 202 ┆ tru ┆ Jo  ┆ 202 ┆ art ┆ htt ┆ ["B ┆ ["L ┆ ["K ┆ 118 ┆ [13 ┆ nyh ┆ 0.9 ┆ Neg │
│ 052 ┆ ser ┆ ble ┆ 3-0 ┆ e   ┆ far ┆ 3-0 ┆ icl ┆ ps: ┆ age ┆ OC" ┆ onf ┆     ┆ 3]  ┆ ede ┆ 934 ┆ ati │
│ 1   ┆ : - ┆ v   ┆ 7-1 ┆     ┆ lig ┆ 7-1 ┆ e_d ┆ //e ┆ rie ┆ ,   ┆ lik ┆     ┆     ┆ r   ┆     ┆ ve  │
│     ┆ Jeg ┆ kal ┆ 2   ┆     ┆ ere ┆ 1   ┆ efa ┆ kst ┆ t", ┆ "LO ┆ t   ┆     ┆     ┆     ┆     ┆     │
│     ┆ gik ┆ dt  ┆ 08: ┆     ┆ opg ┆ 05: ┆ ult ┆ rab ┆ "Ba ┆ C", ┆ og  ┆     ┆     ┆     ┆     ┆     │
│     ┆ i   ┆ 'Kn ┆ 28: ┆     ┆ ave ┆ 05: ┆     ┆ lad ┆ lle ┆ …   ┆ kri ┆     ┆     ┆     ┆     ┆     │
│     ┆ sty ┆ ogl ┆ 14  ┆     ┆ rne ┆ 36  ┆     ┆ et. ┆ rup ┆ "PE ┆ g", ┆     ┆     ┆     ┆     ┆     │
│     ┆ kke ┆ ekn ┆     ┆     ┆ var ┆     ┆     ┆ dk/ ┆ ",  ┆ R"] ┆ "Te ┆     ┆     ┆     ┆     ┆     │
│     ┆ r   ┆ use ┆     ┆     ┆ ,   ┆     ┆     ┆ nyh ┆ …   ┆     ┆ rro ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆ ren ┆     ┆     ┆ jo… ┆     ┆     ┆ ede ┆ "Th ┆     ┆ r"] ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆ …   ┆     ┆     ┆     ┆     ┆     ┆ …   ┆ …   ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│ 980 ┆ Akt ┆ Fle ┆ 202 ┆ fal ┆ Min ┆ 202 ┆ art ┆ htt ┆ ["A ┆ ["P ┆ ["K ┆ 140 ┆ []  ┆ kri ┆ 0.9 ┆ Neg │
│ 360 ┆ ion ┆ re  ┆ 3-0 ┆ se  ┆ dst ┆ 3-0 ┆ icl ┆ ps: ┆ dam ┆ ER" ┆ rim ┆     ┆     ┆ mi  ┆ 948 ┆ ati │
│ 7   ┆ mod ┆ kvi ┆ 6-2 ┆     ┆ ott ┆ 6-0 ┆ e_d ┆ //e ┆ ",  ┆ ,   ┆ ina ┆     ┆     ┆     ┆     ┆ ve  │
│     ┆ svi ┆ nde ┆ 9   ┆     ┆ e   ┆ 8   ┆ efa ┆ kst ┆ "Be ┆ "PE ┆ lit ┆     ┆     ┆     ┆     ┆     │
│     ┆ ndl ┆ r   ┆ 06: ┆     ┆ per ┆ 06: ┆ ult ┆ rab ┆ nte ┆ R", ┆ et" ┆     ┆     ┆     ┆     ┆     │
│     ┆ ere ┆ er  ┆ 49: ┆     ┆ son ┆ 54: ┆     ┆ lad ┆ ",  ┆ …   ┆ ,   ┆     ┆     ┆     ┆     ┆     │
│     ┆ :   ┆ ifø ┆ 26  ┆     ┆ er  ┆ 53  ┆     ┆ et. ┆ …   ┆ "OR ┆ "Be ┆     ┆     ┆     ┆     ┆     │
│     ┆ Sek ┆ lge ┆     ┆     ┆ er  ┆     ┆     ┆ dk/ ┆ "Tw ┆ G"] ┆ dra ┆     ┆     ┆     ┆     ┆     │
│     ┆ s   ┆ pol ┆     ┆     ┆ ble ┆     ┆     ┆ kri ┆ itt ┆     ┆ ger ┆     ┆     ┆     ┆     ┆     │
│     ┆ per ┆ iti ┆     ┆     ┆ vet ┆     ┆     ┆ mi/ ┆ er" ┆     ┆ i"] ┆     ┆     ┆     ┆     ┆     │
│     ┆ …   ┆ …   ┆     ┆     ┆ …   ┆     ┆     ┆ …   ┆ ]   ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│ 980 ┆ Dan ┆ Jul ┆ 202 ┆ fal ┆ Men ┆ 202 ┆ art ┆ htt ┆ ["C ┆ ["P ┆ ["K ┆ 414 ┆ [42 ┆ und ┆ 0.7 ┆ Neg │
│ 352 ┆ sk  ┆ ie  ┆ 3-0 ┆ se  ┆ s   ┆ 3-0 ┆ icl ┆ ps: ┆ oop ┆ ER" ┆ end ┆     ┆ 5]  ┆ erh ┆ 737 ┆ ati │
│ 5   ┆ sku ┆ R.  ┆ 6-2 ┆     ┆ hun ┆ 6-0 ┆ e_d ┆ //e ┆ er" ┆ ,   ┆ t", ┆     ┆     ┆ old ┆     ┆ ve  │
│     ┆ esp ┆ Ølg ┆ 9   ┆   

### Behaviours DataFrame

#### Exploration

This DataFrame contains all information about interactions between users and items over 7 days

In [6]:
dataProcesser.train_behaviors_df.head(1)

shape: (1, 17)
┌─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┐
│ imp ┆ art ┆ imp ┆ rea ┆ scr ┆ dev ┆ art ┆ art ┆ use ┆ is_ ┆ gen ┆ pos ┆ age ┆ is_ ┆ ses ┆ nex ┆ nex │
│ res ┆ icl ┆ res ┆ d_t ┆ oll ┆ ice ┆ icl ┆ icl ┆ r_i ┆ sso ┆ der ┆ tco ┆ --- ┆ sub ┆ sio ┆ t_r ┆ t_s │
│ sio ┆ e_i ┆ sio ┆ ime ┆ _pe ┆ _ty ┆ e_i ┆ e_i ┆ d   ┆ _us ┆ --- ┆ de  ┆ i8  ┆ scr ┆ n_i ┆ ead ┆ cro │
│ n_i ┆ d   ┆ n_t ┆ --- ┆ rce ┆ pe  ┆ ds_ ┆ ds_ ┆ --- ┆ er  ┆ i8  ┆ --- ┆     ┆ ibe ┆ d   ┆ _ti ┆ ll_ │
│ d   ┆ --- ┆ ime ┆ f32 ┆ nta ┆ --- ┆ inv ┆ cli ┆ u32 ┆ --- ┆     ┆ i8  ┆     ┆ r   ┆ --- ┆ me  ┆ per │
│ --- ┆ i32 ┆ --- ┆     ┆ ge  ┆ i8  ┆ iew ┆ cke ┆     ┆ boo ┆     ┆     ┆     ┆ --- ┆ u32 ┆ --- ┆ cen │
│ u32 ┆     ┆ dat ┆     ┆ --- ┆     ┆ --- ┆ d   ┆     ┆ l   ┆     ┆     ┆     ┆ boo ┆     ┆ f32 ┆ tag │
│     ┆     ┆ eti ┆     ┆ f32 ┆     ┆ lis ┆ --- ┆     ┆     ┆     ┆     ┆     ┆ l   ┆     ┆     ┆ e   │
│     ┆     ┆ me[ ┆     ┆     ┆     ┆ t[i ┆ lis ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ --- │
│     ┆     ┆ μs] ┆     ┆     ┆     ┆ 32] ┆ t[i ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ f32 │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆ 32] ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
╞═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╡
│ 149 ┆ nul ┆ 202 ┆ 13. ┆ nul ┆ 2   ┆ [97 ┆ [97 ┆ 139 ┆ fal ┆ nul ┆ nul ┆ nul ┆ fal ┆ 759 ┆ 7.0 ┆ 22. │
│ 474 ┆ l   ┆ 3-0 ┆ 0   ┆ l   ┆     ┆ 786 ┆ 786 ┆ 836 ┆ se  ┆ l   ┆ l   ┆ l   ┆ se  ┆     ┆     ┆ 0   │
│     ┆     ┆ 5-2 ┆     ┆     ┆     ┆ 23, ┆ 57] ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆ 4   ┆     ┆     ┆     ┆ 977 ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆ 07: ┆     ┆     ┆     ┆ 868 ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆ 47: ┆     ┆     ┆     ┆ 2,  ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆ 53  ┆     ┆     ┆     ┆ …   ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆     ┆     ┆     ┆     ┆ 977 ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆     ┆     ┆     ┆     ┆ 872 ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
│     ┆     ┆     ┆     ┆     ┆     ┆ 8]  ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     │
└─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┘

By inspecting null values, we see that of 232887 interactions, around half of article_id's are empty and almost all gender, postcode and age values are lacking. While scroll_percentage also have around half of it's values missing, these may be corresponding to the article_id's that are null

In [7]:
print("Dataframe Shape = ", dataProcesser.train_behaviors_df.shape)
dataProcesser.train_behaviors_df.null_count()

Dataframe Shape =  (232887, 17)


shape: (1, 17)
┌─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┐
│ imp ┆ art ┆ imp ┆ rea ┆ scr ┆ dev ┆ art ┆ art ┆ use ┆ is_ ┆ gen ┆ pos ┆ age ┆ is_ ┆ ses ┆ nex ┆ nex │
│ res ┆ icl ┆ res ┆ d_t ┆ oll ┆ ice ┆ icl ┆ icl ┆ r_i ┆ sso ┆ der ┆ tco ┆ --- ┆ sub ┆ sio ┆ t_r ┆ t_s │
│ sio ┆ e_i ┆ sio ┆ ime ┆ _pe ┆ _ty ┆ e_i ┆ e_i ┆ d   ┆ _us ┆ --- ┆ de  ┆ u32 ┆ scr ┆ n_i ┆ ead ┆ cro │
│ n_i ┆ d   ┆ n_t ┆ --- ┆ rce ┆ pe  ┆ ds_ ┆ ds_ ┆ --- ┆ er  ┆ u32 ┆ --- ┆     ┆ ibe ┆ d   ┆ _ti ┆ ll_ │
│ d   ┆ --- ┆ ime ┆ u32 ┆ nta ┆ --- ┆ inv ┆ cli ┆ u32 ┆ --- ┆     ┆ u32 ┆     ┆ r   ┆ --- ┆ me  ┆ per │
│ --- ┆ u32 ┆ --- ┆     ┆ ge  ┆ u32 ┆ iew ┆ cke ┆     ┆ u32 ┆     ┆     ┆     ┆ --- ┆ u32 ┆ --- ┆ cen │
│ u32 ┆     ┆ u32 ┆     ┆ --- ┆     ┆ --- ┆ d   ┆     ┆     ┆     ┆     ┆     ┆ u32 ┆     ┆ u32 ┆ tag │
│     ┆     ┆     ┆     ┆ u32 ┆     ┆ u32 ┆ --- ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ e   │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆ u32 ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ --- │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ u32 │
╞═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╡
│ 0   ┆ 162 ┆ 0   ┆ 0   ┆ 163 ┆ 0   ┆ 0   ┆ 0   ┆ 0   ┆ 0   ┆ 216 ┆ 228 ┆ 226 ┆ 0   ┆ 0   ┆ 621 ┆ 262 │
│     ┆ 466 ┆     ┆     ┆ 789 ┆     ┆     ┆     ┆     ┆     ┆ 668 ┆ 214 ┆ 546 ┆     ┆     ┆ 8   ┆ 70  │
└─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┘

We see that after removing the interactions from the mainpage, we have been reduced by 60%. Of the remaining 70K interactions, only 2% lack scroll data, this can be replaced by using mean. Furthermore, we see that around >92% of gender, postcode and age values are still missing, so we can't really predict these

In [8]:
df = dataProcesser.process_dataframe(dataProcesser.train_behaviors_df, filter_null_columns=["article_id"])
print("Dataframe Shape = ", df.shape)
df.null_count()

Dataframe Shape =  (70421, 17)


shape: (1, 17)
┌─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┐
│ imp ┆ art ┆ imp ┆ rea ┆ scr ┆ dev ┆ art ┆ art ┆ use ┆ is_ ┆ gen ┆ pos ┆ age ┆ is_ ┆ ses ┆ nex ┆ nex │
│ res ┆ icl ┆ res ┆ d_t ┆ oll ┆ ice ┆ icl ┆ icl ┆ r_i ┆ sso ┆ der ┆ tco ┆ --- ┆ sub ┆ sio ┆ t_r ┆ t_s │
│ sio ┆ e_i ┆ sio ┆ ime ┆ _pe ┆ _ty ┆ e_i ┆ e_i ┆ d   ┆ _us ┆ --- ┆ de  ┆ u32 ┆ scr ┆ n_i ┆ ead ┆ cro │
│ n_i ┆ d   ┆ n_t ┆ --- ┆ rce ┆ pe  ┆ ds_ ┆ ds_ ┆ --- ┆ er  ┆ u32 ┆ --- ┆     ┆ ibe ┆ d   ┆ _ti ┆ ll_ │
│ d   ┆ --- ┆ ime ┆ u32 ┆ nta ┆ --- ┆ inv ┆ cli ┆ u32 ┆ --- ┆     ┆ u32 ┆     ┆ r   ┆ --- ┆ me  ┆ per │
│ --- ┆ u32 ┆ --- ┆     ┆ ge  ┆ u32 ┆ iew ┆ cke ┆     ┆ u32 ┆     ┆     ┆     ┆ --- ┆ u32 ┆ --- ┆ cen │
│ u32 ┆     ┆ u32 ┆     ┆ --- ┆     ┆ --- ┆ d   ┆     ┆     ┆     ┆     ┆     ┆ u32 ┆     ┆ u32 ┆ tag │
│     ┆     ┆     ┆     ┆ u32 ┆     ┆ u32 ┆ --- ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ e   │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆ u32 ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ --- │
│     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆     ┆ u32 │
╞═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╡
│ 0   ┆ 0   ┆ 0   ┆ 0   ┆ 178 ┆ 0   ┆ 0   ┆ 0   ┆ 0   ┆ 0   ┆ 654 ┆ 687 ┆ 679 ┆ 0   ┆ 0   ┆ 317 ┆ 912 │
│     ┆     ┆     ┆     ┆ 0   ┆     ┆     ┆     ┆     ┆     ┆ 81  ┆ 85  ┆ 72  ┆     ┆     ┆ 3   ┆ 9   │
└─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┘

#### Summary

**Filtering:** We can remove gender, postcode and age due to lacking data. We canalso remove next_read_time, next_scroll_percentage, article_ids_inview, article_ids_clicked as future data won't be used in our predictive algorithms. 

**Sorting:** We can sort by published_time if we want to. But this isn't required

**Non-Null:** We can sort out all rows where article_id is null

**Predict:** We can predict all scroll_percentage left after article_id is filtered out

In [10]:
remove_col = ["gender", "postcode", "age", "next_read_time", "next_scroll_percentage", "article_ids_inview", "article_ids_clicked"]
non_null = ["article_id"]
predict = ["scroll_percentage"] # The standard method is mean, which will work well 

dataProcesser.process_train_test_df(train_df=dataProcesser.train_behaviors_df, test_df=dataProcesser.test_behaviors_df, remove_columns = remove_col, filter_null_columns = non_null, predict_columns = predict).head()

shape: (5, 10)
┌─────────┬─────────┬─────────┬─────────┬─────────┬─────────┬─────────┬─────────┬─────────┬────────┐
│ impress ┆ article ┆ impress ┆ read_ti ┆ scroll_ ┆ device_ ┆ user_id ┆ is_sso_ ┆ is_subs ┆ sessio │
│ ion_id  ┆ _id     ┆ ion_tim ┆ me      ┆ percent ┆ type    ┆ ---     ┆ user    ┆ criber  ┆ n_id   │
│ ---     ┆ ---     ┆ e       ┆ ---     ┆ age     ┆ ---     ┆ u32     ┆ ---     ┆ ---     ┆ ---    │
│ u32     ┆ i32     ┆ ---     ┆ f32     ┆ ---     ┆ i8      ┆         ┆ bool    ┆ bool    ┆ u32    │
│         ┆         ┆ datetim ┆         ┆ f32     ┆         ┆         ┆         ┆         ┆        │
│         ┆         ┆ e[μs]   ┆         ┆         ┆         ┆         ┆         ┆         ┆        │
╞═════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╪════════╡
│ 153068  ┆ 9778682 ┆ 2023-05 ┆ 78.0    ┆ 100.0   ┆ 1       ┆ 151570  ┆ false   ┆ false   ┆ 1976   │
│         ┆         ┆ -24 07: ┆         ┆         ┆         ┆         ┆         ┆         ┆        │
│         ┆         ┆ 09:04   ┆         ┆         ┆         ┆         ┆         ┆         ┆        │
│ 153070  ┆ 9777492 ┆ 2023-05 ┆ 26.0    ┆ 100.0   ┆ 1       ┆ 151570  ┆ false   ┆ false   ┆ 1976   │
│         ┆         ┆ -24 07: ┆         ┆         ┆         ┆         ┆         ┆         ┆        │
│         ┆         ┆ 13:14   ┆         ┆         ┆         ┆         ┆         ┆         ┆        │
│ 153071  ┆ 9778623 ┆ 2023-05 ┆ 125.0   ┆ 100.0   ┆ 1       ┆ 151570  ┆ false   ┆ false   ┆ 1976   │
│         ┆         ┆ -24 07: ┆         ┆         ┆         ┆         ┆         ┆         ┆        │
│         ┆         ┆ 11:08   ┆         ┆         ┆         ┆         ┆         ┆         ┆        │
│ 153075  ┆ 9777492 ┆ 2023-05 ┆ 26.0    ┆ 100.0   ┆ 1       ┆ 151570  ┆ false   ┆ false   ┆ 1976   │
│         ┆         ┆ -24 07: ┆         ┆         ┆         ┆         ┆         ┆         ┆        │
│         ┆         ┆ 13:58   ┆         ┆         ┆         ┆         ┆         ┆         ┆        │
│ 153078  ┆ 9777492 ┆ 2023-05 ┆ 7.0     ┆ 100.0   ┆ 1       ┆ 151570  ┆ false   ┆ false   ┆ 1976   │
│         ┆         ┆ -24 07: ┆         ┆         ┆         ┆         ┆         ┆         ┆        │
│         ┆         ┆ 13:46   ┆         ┆         ┆         ┆         ┆         ┆         ┆        │
└─────────┴─────────┴─────────┴─────────┴─────────┴─────────┴─────────┴─────────┴─────────┴────────┘

### Document Vectors

#### Exploration

We see that the dataProcesser contains no empty fields. So there isn't really much to do with this vector

In [12]:
dataProcesser.document_vectors_df.null_count()

shape: (1, 2)
┌────────────┬─────────────────┐
│ article_id ┆ document_vector │
│ ---        ┆ ---             │
│ u32        ┆ u32             │
╞════════════╪═════════════════╡
│ 0          ┆ 0               │
└────────────┴─────────────────┘

#### Summary

This vector is perfect just the way it is. It could however be combined with the articles dataframe if needed